zzre/game/systems/dialog/DialogChestPuzzle.cs

```diff
@@ -0,0 +1,209 @@
+                var button = preload.CreateButton(entity)
+                    .With(IDCell)
+                    .With(offset + new Vector2(46 * col, 46 * row))
+                    .With(new components.ui.ButtonTiles(puzzle.BoardState[cell] ? 1 : 2))
+                    .With(preload.Swt000)
+                    .Build();
+                button.Set(button.Get<Rect>().GrownBy(new Vector2(1, 1))); // No gaps
+                button.Set(new components.ui.Silent());
+            }
+        }
+
+        return entity;
+    }
+
+    private readonly (int row, int col)[] flipped = [
+        (0, 0),
+        (-1, 0),
+        (0, -1),
+        (1, 0),
+        (0, 1)
+    ];
+
+    private void UpdateBoard(DefaultEcs.Entity parent, ref components.DialogChestPuzzle puzzle, int cellId)
+    {
+        puzzle.NumAttempts += 1;
+        World.Publish(new messages.SpawnSample("resources/audio/sfx/gui/_g000.wav"));
+
+        int row = cellId / puzzle.Size;
+        int col = cellId % puzzle.Size;
+
+        foreach (var coord in flipped)
+        {
+            if (coord.row + row < puzzle.Size && coord.row + row >= 0 &&
+                coord.col + col < puzzle.Size && coord.col + col >= 0)
+            {
+                var cell = (coord.row + row) * puzzle.Size + (coord.col + col);
+                puzzle.BoardState[cell] = !puzzle.BoardState[cell];
+            }
+        }
+
+        if (puzzle.BoardState.All(x => x) || puzzle.BoardState.All(x => !x))
+            Succeed(parent, ref puzzle);
+
+        puzzle.Board.Dispose();
+        puzzle.Board = CreateBoard(parent, ref puzzle);
+        puzzle.Attempts.Dispose();
+        puzzle.Attempts = CreateAttempts(parent, ref puzzle);
+    }
+
+    private void Succeed(DefaultEcs.Entity parent, ref components.DialogChestPuzzle puzzle)
+    {
+        World.Publish(new messages.SpawnSample($"resources/audio/sfx/specials/_s022.wav"));
+
+        preload.CreateLabel(parent)
+            .With(new Vector2(0, -126))
+            .With(components.ui.FullAlignment.Center)
+            .With(preload.Fnt001)
+            .WithText(db.GetText(UIDChestOpened).Text)
+            .Build();
+
+        if (savegame.switchGameMinMoves > puzzle.NumAttempts)
+            savegame.switchGameMinMoves = puzzle.NumAttempts;
+
+        puzzle.LockBoard = true;
+
+        puzzle.Action.Dispose();
+        puzzle.Action = preload.CreateSingleDialogButton(parent, UIDNext, IDNext, puzzle.BgRect, buttonOffsetY: -45f);
+    }
+
+    private void HandleElementDown(DefaultEcs.Entity entity, components.ui.ElementId clickedId)
+    {
+        var uiEntity = Set.GetEntities()[0];
+        ref var puzzle = ref uiEntity.Get<components.DialogChestPuzzle>();
+        ref var script = ref puzzle.DialogEntity.Get<components.ScriptExecution>();
+
+        if (!puzzle.LockBoard && clickedId.InRange(new components.ui.ElementId(0), new components.ui.ElementId(puzzle.Size * puzzle.Size), out var cellId)) {
```

Comment:
this line would be shorter with a `FirstCellId` static readonly variable, then the second one can be `FirstCellId + size*size`